numpy/core/src/multiarray/arrayobject.c

```diff
@@ -85,17 +85,56 @@ PyArray_Size(PyObject *op)
 NPY_NO_EXPORT int
 PyArray_SetUpdateIfCopyBase(PyArrayObject *arr, PyArrayObject *base)
 {
+    int ret;
+#ifdef PYPY_VERSION
+  #ifndef DEPRECATE_UPDATEIFCOPY
+    #define DEPRECATE_UPDATEIFCOPY
+  #endif
+#endif
+
```

Comment:
This seems odd to me. Why are we only deprecating this for PyPy? And IIUC, `DEPRECATE` should always be called.

Reply:
See the design [comment](https://github.com/numpy/numpy/pull/9639#issue-254637365)  at the top of the pull request, and issue #9714 which covers the problem in depth. We cannot yet issue this warning because it will be emitted by ``nditer( ..., [updateifcopy], ...)``, with no option to silence it. Once this pull request lands, I will turn my attention to solving issue #9714, hopefully by choosing one of the alternatives mentioned there

Reply:
Either the warning can be issued or it cannot. If it cannot, then surely pypy will have the same problem?

Reply:
Ordinarily, I would say that you should rename the function so that internal users can use it without emitting the warning but external users cannot. But that may increase merge conflicts, and I know what it's like to be worried about having to rebase.

Reply:
The PR also moves all the internal users it can over to the new API, so there's no need for a no-warning version of the old API.

Reply:
Until issue #9714 is resolved, ``npyiter_allocate_arrays()`` in ``nditer_constr.c`` still uses the old interface (i.e. calls this function), so yes, there is still a need for a no-warning version of the old API on CPython

It is hard enough to make this enormous change in NumPy without adding in yet another dimension of pain in fixing #9714, which will require [engaging with the community](http://numpy-discussion.10968.n7.nabble.com/nditer-and-updateifcopy-semantics-advice-needed-td44739.html) to try to determine if and how to fix it 

However, IMO in the mean time, users of PyPy should be warned that using the semantics in #9714 on PyPy is not well supported, and could cause wrong answers. It seems this compromise is the least evil of all the alternatives.

Reply:
> It is hard enough to make this enormous change in NumPy without adding in yet another dimension of pain in fixing #9714, which will require engaging with the community to try to determine if and how to fix it

+1. If this requires additional refactoring to avoid the warning, I don't think you should do it for now.

> However, IMO in the mean time, users of PyPy should be warned that using the semantics in #9714 on PyPy is not well supported, and could cause wrong answers. It seems this compromise is the least evil of all the alternatives.

I disagree. I think we should leave the warning as a comment for now because there are still internal users. I don't think numpy should try to behave differently for a different implementation.

Reply:
@xoviat: also logic on warning on PyPy first is that right now, the remaining `nditer` case gives incorrect results on PyPy but not on CPython. So we don't want to hassle people on CPython until they have an alternative to move to, but on PyPy you probably want to know your code is broken even if there isn't a clear alternative.

Reply:
I supposed it's fine as it is then.